symcache/src/writer.rs

```diff
@@ -216,35 +214,41 @@ where
         writer.set_arch(object.arch());
         writer.set_debug_id(object.debug_id());
 
-        let mut last_address = 0;
-        let mut symbols = object.symbol_map().into_iter().peekable();
         let session = object
             .debug_session()
             .context(SymCacheErrorKind::BadDebugFile)?;
 
         for function in session.functions() {
             let function = function.context(SymCacheErrorKind::BadDebugFile)?;
+            writer.add_function(function)?;
+        }
+
+        // Sort the files to efficiently add symbols from the symbol table in linear time
+        // complexity. When the writer finishes, it will sort again with the added symbols.
+        writer.ensure_sorted();
```

Comment:
This is where the magic happens:
1. We insert functions in the original order. No point in assuming they are sorted.
2. Sort them
3. Do one pass through functions and symbols and add the symbols
4. Flush the writer, which will sort again and do its thing.